Multiple middlewares corrupt TS typings.

**Describe the bug**

Adding middlewares alters the return type of `step.run`.

**To Reproduce**

I made a minimal repro in a small repository
* Repro steps: https://github.com/PatrickNaushaPlanoly/inngest-typing-problems#reproduce
* Relevant line of code: https://github.com/PatrickNaushaPlanoly/inngest-typing-problems/blob/76ab6ef8e7ef7377496eecb60ec8e897ffda9b70/src/repro.ts#L70

**Expected behavior**

The `middleware` passed into `Inngest` would not affect the type generics used by `step.run`.

**Actual behavior**

See inline comments about compilation errors below.

```ts
class MiddlewareA extends Middleware.BaseMiddleware {
  readonly id = "a";
}
class MiddlewareB extends Middleware.BaseMiddleware {
  readonly id = "b";
}

// Compiles fine:
const client1 = new Inngest({ id: "client-1mw", middleware: [MiddlewareA] });
client1.createFunction({ id: "f1", triggers: [{ event: "demo/event" }] }, async ({ step }) => {
  const widget = await step.run("load", async () => widgetValue);
  const mediaId: string = widget.media[0].mediaId; // OK
  return mediaId;
});

// Collapses the type of media[0] below to just 'JsonifyObject<{}>'
// such that the `mediaId` property does not exist.
const client2 = new Inngest({
  id: "client-2mw",
  // 👇👇
  middleware: [MiddlewareA, MiddlewareB], // 📢 This is the only meaningful difference between `client1` and `client2`
  // 👆👆
});
client2.createFunction({ id: "f2", triggers: [{ event: "demo/event" }] }, async ({ step }) => {
  const widget = await step.run("load", async () => widgetValue);
  // error TS2339: Property 'mediaId' does not exist on type 'JsonifyObject<{}>':
  const mediaId: string = widget.media[0].mediaId; // <- TS2339 here
  return mediaId;
});
```

**Screenshots / Stack trace dump**

<img width="613" height="115" alt="Image" src="https://github.com/user-attachments/assets/fb476755-114f-431b-a2ca-2c18713137fa" />
<img width="926" height="211" alt="Image" src="https://github.com/user-attachments/assets/92b22f8f-2283-4acf-bcde-1fb9be91ee0a" />

**System info**
- typescript 6.0.3
- inngest: 4.7.0

--- a/.changeset/full-dingos-brush.md
+++ b/.changeset/full-dingos-brush.md
@@ -1,0 +1,5 @@
+---
+"inngest": patch
+---
+
+Fix `step.run` and `step.invoke` output types collapsing to `{}` for objects with optional properties when using two or more middleware

--- a/packages/inngest/src/components/middleware/middleware.test.ts
+++ b/packages/inngest/src/components/middleware/middleware.test.ts
@@ -33,3 +33,61 @@ test("stepOutputTransform does not affect step.invoke return type", () => {
     },
   );
 });
+
+// stacked default output transforms apply `Jsonify` repeatedly,
+// which used to collapse objects with optional properties to `{}`.
+test("multiple middleware preserve step.run output with optional properties", () => {
+  class MiddlewareA extends Middleware.BaseMiddleware {
+    readonly id = "a";
+  }
+  class MiddlewareB extends Middleware.BaseMiddleware {
+    readonly id = "b";
+  }
+
+  type Widget = { media: { mediaId: string; label?: string }[] };
+
+  const client = new Inngest({
+    id: "test",
+    middleware: [MiddlewareA, MiddlewareB],
+  });
+
+  client.createFunction(
+    { id: "fn", triggers: [{ event: "any" }] },
+    async ({ step }) => {
+      const widget = await step.run(
+        "load",
+        async (): Promise<Widget> => ({ media: [] }),
+      );
+
+      expectTypeOf(widget.media).not.toBeAny();
+      expectTypeOf(widget.media).toEqualTypeOf<
+        { mediaId: string; label?: string }[]
+      >();
+      expectTypeOf(widget.media[0]!.mediaId).toEqualTypeOf<string>();
+    },
+  );
+});
+
+// a handler returning a `step.run` result already has `Jsonify` in its return type,
+// so `step.invoke` applies it a second time.
+test("step.invoke preserves optional properties in invoked step.run output", () => {
+  type Widget = { media: { mediaId: string; label?: string }[] };
+
+  const client = new Inngest({ id: "test" });
+
+  const childFn = client.createFunction(
+    { id: "child", triggers: [{ event: "any" }] },
+    async ({ step }) =>
+      step.run("load", async (): Promise<Widget> => ({ media: [] })),
+  );
+
+  client.createFunction(
+    { id: "parent", triggers: [{ event: "any" }] },
+    async ({ step }) => {
+      const widget = await step.invoke("invoke-child", { function: childFn });
+
+      expectTypeOf(widget.media).not.toBeAny();
+      expectTypeOf(widget.media[0]!.mediaId).toEqualTypeOf<string>();
+    },
+  );
+});

--- a/packages/inngest/src/helpers/jsonify.test.ts
+++ b/packages/inngest/src/helpers/jsonify.test.ts
@@ -108,6 +108,35 @@ describe("Jsonify", () => {
     });
   });
 
+  // The property-access assertions are the real guards here: the `IsEqual`
+  // assertions pass even on broken variants because the deferred aliases
+  // compare equal before resolution.
+  describe("#1532", () => {
+    test("re-applying Jsonify to an object with optional properties is a no-op", () => {
+      type Widget = { media: { mediaId: string; label?: string }[] };
+
+      type Once = Jsonify<Widget>;
+      type Twice = Jsonify<Once>;
+
+      assertType<IsEqual<Once, Twice>>(true);
+      assertType<IsEqual<Twice["media"][number]["mediaId"], string>>(true);
+    });
+
+    test("re-applying Jsonify to a mapped type with optional properties is a no-op", () => {
+      interface Foo {
+        // biome-ignore lint/suspicious/noExplicitAny: intentional
+        [x: string]: any;
+        items: { id: string; label?: string }[];
+      }
+
+      type Once = Jsonify<Foo>;
+      type Twice = Jsonify<Once>;
+
+      assertType<IsEqual<Once, Twice>>(true);
+      assertType<IsEqual<Twice["items"][number]["id"], string>>(true);
+    });
+  });
+
   describe("#537", () => {
     describe("nested { name: string; } object is preserved", () => {
       test("when nullable", () => {

--- a/packages/inngest/src/helpers/jsonify.ts
+++ b/packages/inngest/src/helpers/jsonify.ts
@@ -3,13 +3,10 @@
  * to represent to users how a value will be serialized and deserialized as it
  * passes to and from an Inngest Server.
  *
- * We do not use the `type-fest` package directly due to some version
- * compatibility issues:
- *
- * - `inngest` supports `typescript@>=4.7`
- * - `type-fest@4` supports `typescript@>=5.1`, so the maximum version we can
- *   use is `type-fest@3`
- * - `type-fest@3` is not compatible with `typescript@5.4`
+ * We do not use the `type-fest` package directly because this copy
+ * intentionally diverges from it: `unknown` is preserved, objects mixing
+ * index signatures with known keys are handled, and re-applying `Jsonify` to
+ * its own output is a no-op.
  */
 import type {
   IsAny,
@@ -37,9 +34,17 @@ type JsonifyList<T extends UnknownArray> = T extends readonly []
       ? []
       : Array<T[number] extends NotJsonable ? null : Jsonify<T[number]>>;
 
-type FilterJsonableKeys<T extends object> = {
-  [Key in keyof T]: T[Key] extends NotJsonable ? never : Key;
-}[keyof T];
+/**
+Diverges from `type-fest`: optional properties leak `undefined` into the key
+union, which collapses objects to `{}` when `Jsonify` is re-applied to its
+own output.
+*/
+type FilterJsonableKeys<T extends object> = Exclude<
+  {
+    [Key in keyof T]: T[Key] extends NotJsonable ? never : Key;
+  }[keyof T],
+  undefined
+>;
 
 /**
 JSON serialize objects (not including arrays) and classes.

